### PR TITLE
week5 - lab revised version

### PR DESCRIPTION
**Pull request by AngelaZZZ-611**
changed all the edits mentioned: replaced all screenshots figures to code, expanded the custom_grads part, re-drew all figures instead of screenshots. Also, update 05.md and _config.yml files.

### CODE REVIEW

File: docs/chapters/05-3.md [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+```
+
+### 1D Convolution
+
+We have $1$ dimensional convolution going from $2$ channels (stereophonic signal) to $16$ channels ($16$ kernels) with kernel size of $3$ and stride of $1$. We then have $16$ kernels with thickness $2$ and length $3$. Let's assume that the input signal has a batch of size $1$ (one signal),  $2$ channels and $64$ samples. The resulting output layer has one signal, $16$ channels and the length of the signal is $62$ ($=64-3+1$). Also, if we output the bias size, we'll find the bias size is $16$, since we have one bias per weight. 
+
+```python
+import torch
+from torch import nn
+
```
Comment (Atcold) [inline docs/chapters/05-3.md]: Not necessary.
```suggestion
import torch
from torch import nn

```

File: docs/chapters/05-3.md [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+conv = nn.Conv1d(2, 16, 3)
```
Comment (Atcold) [inline docs/chapters/05-3.md]: Explain what `2`, `16`, and `3` stand for.

File: docs/chapters/05-3.md [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    z.backward()  # PyTorch will throw an error here, since z has no grad accum.
+except RuntimeError as e:
+    print('RuntimeError!!! >:[')
+    print(e)
+```
+
+
+## More Stuff -- Custom Grads 
+
+Also, instead of basic numerical operations, we can generate our own self-defined modules/functions, which can be plugged into the nerual graph. (see *notebook -- extra -- b-custom_grads.ipynb*)
```
Comment (Atcold) [inline docs/chapters/05-3.md]: Add link to notebook.

File: docs/chapters/05-3.md [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+&a = \frac{1}{4}\sum_{i=1}^{4}z_i\\[2mm]
+&z_i = 3y_i^2 = 3(x_i-2)^2\\[2mm]
+&\frac{da}{dx_i} = \frac{1}{4}\times3\times2(x_i-2) = \frac{3}{2}x_i-3\\[2mm]
+&x = \begin{bmatrix} 1&2\\3&4\end{bmatrix}
+\\[2mm] &(\frac{da}{dx_i})^{T} = \begin{bmatrix} 1.5-3&3-3\\[2mm]4.5-3&6-3\end{bmatrix}=\begin{bmatrix} -1.5&0\\[2mm]1.5&3\end{bmatrix}
+\end{aligned}
+$$
+
+Whenever you use partial derivative in PyTorch, you get the same shape of the original data. But the correct Jacobian thing should be the transpose.
+
```
Comment (Atcold) [inline docs/chapters/05-3.md]: One extra empty line.
```suggestion


```

File: docs/chapters/05-3.md [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+x = torch.rand(1, 20, 64, 128)
+conv = nn.Conv2d(20, 16, (3, 5))
+conv.weight.size()	# output: torch.Size([16, 20, 3, 5])
+conv(x).size()	# output: torch.Size([1, 16, 62, 124])
+```
+
+If we want to achieve the same dimensionality, we can have paddings. Continuing the code above, we can add new parameters to the convolution function: `stride=1` and `padding=(1, 2)` , which means $1$ on $y$ direction ($1$ at the top and $1$ at the bottom) and $2$ on $x$ direction. Then the output signal is in the same size compared to the input signal. The number of dimensions that is required to store the collection of kernels when you perform $2D$ convolution is $4$.
+
+```python
+conv = nn.Conv2d(20, 16, (3, 5), 1, (1, 2))
```
Comment (Atcold) [inline docs/chapters/05-3.md]: Add a comment where the numbers are explained.

File: docs/chapters/05-3.md [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+conv = nn.Conv1d(2, 16 , 5)
+conv(x).size()	# output: torch.Size([1, 16, 60])
+```
+
+### 2D Convolution
+
+We first define the input data as one sample, $20$ channels with height $64$ and width $128$. $2D$ convolution has $20$ channels from input and $16$ kernels with size of $3 \times 5$. After convolution, the output data has one sample, $16$ channels with height $62 (=64-3+1)$ and width $124 (=128-5+1)$. 
+
+```python
+x = torch.rand(1, 20, 64, 128)
```
Comment (Atcold) [inline docs/chapters/05-3.md]: Explain what these numbers are in a code comment.

File: docs/chapters/05-3.md [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+nn.Conv1d(
+	in_channels,
+	out_channels,
+	kernel_size,
+	stride=1,
+	padding=0,
+	dilation=1,
+	groups=1,
+	bias=True,
+	padding_mode='zeros',
```
Comment (Atcold) [inline docs/chapters/05-3.md]: Add `  # explanation` for each line. That is two spaces and a #, followed by a space and an explanation (without capital letter at the beginning).

File: docs/chapters/05-3.md [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+{a_{m 2}}
+\end{array}\right] x_{2}+\ldots \ldots+\left[\begin{array}{c}
+{a_{1 n}} \\
+{a_{2 n}} \\
+{\vdots} \\
+{a_{m n}}
+\end{array}\right] x_{n}
+$$
+
+This can be viewed as weighted sum of the columns of matrix $A$.
```
Comment (Atcold) [inline docs/chapters/05-3.md]: We covered this in the previous lab.
See https://atcold.github.io/pytorch-Deep-Learning/chapters/04-1/.
Perhaps it's not worth repeating ourselves on the website?
